File: plugins/claude-agent-app-builder/templates/python/custom_tool.py
"""
Custom Tool Example - Weather Tool
Demonstrates how to create a custom tool for Claude Agent SDK
"""

import asyncio
import json
from typing import Any
from claude_agent_sdk import tool, create_sdk_mcp_server, ClaudeSDKClient, ClaudeAgentOptions


# Define a custom tool using @tool decorator
@tool(
    name="get-weather",
    description="Get current weather information for a specific location",
    input_schema={
        "type": "object",
        "properties": {
            "location": {
                "type": "string",
                "description": "City name or ZIP code"
            },
            "units": {
                "type": "string",
                "enum": ["celsius", "fahrenheit"],
                "description": "Temperature units",
                "default": "celsius"
            }
        },
        "required": ["location"]
    }
)
async def get_weather(args: dict[str, Any]) -> dict[str, Any]:
    """
    Get weather information for a location
    In a real implementation, this would call a weather API
    """
    location = args["location"]
    units = args.get("units", "celsius")

    # Simulated weather data (replace with actual API call)
    weather_data = {
        "location": location,
        "temperature": 22 if units == "celsius" else 72,
        "units": units,
        "conditions": "Partly cloudy",
        "humidity": 65,
        "wind_speed": 15
    }

    return {
        "content": [{
            "type": "text",
            "text": json.dumps(weather_data, indent=2)
        }]
    }


# Another example tool - Calculator
@tool(
    name="calculate",
    description="Perform mathematical calculations",
    input_schema={
        "type": "object",
        "properties": {
            "expression": {
                "type": "string",
                "description": "Mathematical expression to evaluate (e.g., '2 + 2', '10 * 5')"
            }
        },
        "required": ["expression"]
    }
)
async def calculate(args: dict[str, Any]) -> dict[str, Any]:
    """
    Safely evaluate mathematical expressions
    """
    expression = args["expression"]

    try:
        # WARNING: eval() is dangerous! Use a proper math parser in production
        # This is just for demonstration purposes
        result = eval(expression, {"__builtins__": {}}, {})

        return {
            "content": [{
                "type": "text",
                "text": f"Result: {result}"
            }]
        }
    except Exception as e:
        return {
            "content": [{
                "type": "text",
                "text": f"Error: {str(e)}"
            }],
            "isError": True
        }


# Create MCP server with custom tools
def create_tools_server():
    """Create an MCP server with custom tools"""
    return create_sdk_mcp_server(
        name="custom-tools",
        tools=[get_weather, calculate]
    )


# Example: Using the custom tools in an agent
async def main():
    """Demonstrate using custom tools in an agent"""

    # Create the MCP server
    tools_server = create_tools_server()

    # Configure agent with custom tools
    options = ClaudeAgentOptions(
        model="claude-sonnet-4-20250514",
        system_prompt="You are a helpful assistant with access to weather and calculator tools.",
        permission_mode="acceptEdits",
        mcp_servers={
            "custom-tools": {
                "server": tools_server
            }
        }
    )

    async with ClaudeSDKClient(options=options) as client:
        # Example queries that use the custom tools
        queries = [
            "What's the weather like in Tokyo?",
            "Calculate 25 * 4 + 10",
            "Convert the temperature in Paris to Fahrenheit"
        ]

        for query in queries:
            print(f"\n🔍 Query: {query}")

            await client.query(query)

            async for message in client.receive_response():
                if message.type == "assistant":
                    for content in message.content:
                        if content["type"] == "text":
                            print(f"🤖 Response: {content['text']}")

                elif message.type == "result":
                    if message.error:
                        print(f"❌ Error: {message.error}")
                    else:
                        print("✅ Completed")


if __name__ == "__main__":
    print("Custom Tool Example - Weather and Calculator\n")
    asyncio.run(main())

File: plugins/claude-agent-app-builder/templates/typescript/custom-tool.ts
/**
 * Custom Tool Example - Weather Tool
 * Demonstrates how to create custom tools for Claude Agent SDK
 */

import { tool, createSdkMcpServer, ClaudeSDKClient } from '@anthropic-ai/claude-agent-sdk';
import { z } from 'zod';

/**
 * Weather Tool - Get current weather information
 */
const weatherTool = tool(
  'get-weather',
  'Get current weather information for a specific location',
  z.object({
    location: z.string().describe('City name or ZIP code'),
    units: z.enum(['celsius', 'fahrenheit']).optional().describe('Temperature units'),
  }),
  async (args) => {
    // Simulated weather data (replace with actual API call)
    const weatherData = {
      location: args.location,
      temperature: args.units === 'fahrenheit' ? 72 : 22,
      units: args.units ?? 'celsius',
      conditions: 'Partly cloudy',
      humidity: 65,
      wind_speed: 15,
    };

    return {
      content: [
        {
          type: 'text' as const,
          text: JSON.stringify(weatherData, null, 2),
        },
      ],
    };
  }
);

/**
 * Calculator Tool - Perform mathematical calculations
 */
const calculatorTool = tool(
  'calculate',
  'Perform mathematical calculations',
  z.object({
    expression: z.string().describe("Mathematical expression to evaluate (e.g., '2 + 2', '10 * 5')"),
  }),
  async (args) => {
    try {
      // WARNING: eval() is dangerous! Use a proper math parser in production
      // This is just for demonstration purposes
      const result = Function(`"use strict"; return (${args.expression})`)();

      return {
        content: [
          {
            type: 'text' as const,
            text: `Result: ${result}`,
          },
        ],
      };
    } catch (error) {
      return {
        content: [
          {
            type: 'text' as const,
            text: `Error: ${error instanceof Error ? error.message : 'Unknown error'}`,
          },
        ],
        isError: true,
      };
    }
  }
);

/**
 * Create MCP server with custom tools
 */
function createToolsServer() {
  return createSdkMcpServer({
    name: 'custom-tools',
    version: '1.0.0',
    tools: [weatherTool, calculatorTool],
  });
}

/**
 * Example: Using the custom tools in an agent
 */
async function main() {
  console.log('Custom Tool Example - Weather and Calculator\n');

  // Create the MCP server
  const toolsServer = createToolsServer();

  // Configure agent with custom tools
  const client = new ClaudeSDKClient({
    model: 'claude-sonnet-4-20250514',
    systemPrompt: 'You are a helpful assistant with access to weather and calculator tools.',
    permissionMode: 'acceptEdits',
    mcpServers: {
      'custom-tools': {
        server: toolsServer,
      },
    },
  });

  try {
    // Example queries that use the custom tools
    const queries = [
      "What's the weather like in Tokyo?",
      'Calculate 25 * 4 + 10',
      'Convert the temperature in Paris to Fahrenheit',
    ];

    for (const query of queries) {
      console.log(`\n🔍 Query: ${query}`);

      await client.query(query);

      for await (const message of client.receiveResponse()) {
        if (message.type === 'assistant') {
          for (const content of message.content) {
            if (content.type === 'text') {
              console.log(`🤖 Response: ${content.text}`);
            }
          }
        } else if (message.type === 'result') {
          if (message.error) {
            console.error(`❌ Error: ${message.error}`);
          } else {
            console.log('✅ Completed');
          }
        }
      }
    }
  } finally {
    await client.close();
  }
}

// Run the example
main().catch((error) => {
  console.error('Fatal error:', error);
  process.exit(1);
});
